### PR TITLE
Prevent Uncaught Exception from `ThinEngine.runRenderLoop` when `window.SetTimeout` in not defined

### PR DESCRIPTION
**Pull request by barroij**
I am having an uncaught exception when using BaylonJs in a NodeJs environment because of a call to `window.setTimeout`.

It is not very clear in the code what is the standard, as sometime we call the global `setTimeout` and sometime we call `window.setTimeout`. 

My take on this is that it should be ok to always call the globals `setTimeout`/ `clearTimeout` as they are bound to be defined and are the very same functions as `window.setTimeout`/ `window.clearTimeout` in most cases.

Though this PR do not change them all. I just make some 2 files more consistent where we used to use a mix of both, and I updated the `ThinEngine.QueueNewFrame` to NOT call `window.setTimeout`, but to fall back on the global `setTimeout`.

To really be "environment  independant", we should never call a function that is "environment  dependant" directly, and put them all behind a configurable callback registry. This registry would be initialized with default values, and could be customizes by the user depending on its needs. But I'm not sure how it is doable, and anyway that's another story :)

### CODE REVIEW

File: packages/dev/core/src/Audio/sound.ts [hunk 1]
```diff
@@ -355,7 +355,7 @@ export class Sound {
                             this._isReadyToPlay = true;
                             // Simulating a ready to play event to avoid breaking code path
                             if (this._readyToPlayCallback) {
-                                window.setTimeout(() => {
+                                setTimeout(() => {
```
Comment (barroij) [inline packages/dev/core/src/Audio/sound.ts:358]: In this file, `window.setTimeout` was used 4 times, and the global `setTimeout` once.
I changed it to always use the global `setTimeout`

Note that modifications in this file are not related on the issue I was encountering.

File: packages/dev/inspector/src/components/actionTabs/tabs/propertyGrids/materials/textures/textureEditorComponent.tsx [hunk 2]
```diff
@@ -200,7 +200,7 @@ export class TextureEditorComponent extends React.Component<ITextureEditorCompon
 
     textureDidUpdate() {
         if (this._timer != null) {
-            clearTimeout(this._timer);
+            window.clearTimeout(this._timer);
```
Comment (barroij) [inline packages/dev/inspector/src/components/actionTabs/tabs/propertyGrids/materials/textures/textureEditorComponent.tsx:203]: Here I use `window.clearTimeout` instead of the global `clearTimeout` for consistency, as we call `window.setTimeout` 2 lines below.

File: packages/dev/core/src/Engines/thinEngine.ts [hunk 3]
```diff
@@ -5852,31 +5852,34 @@ export class ThinEngine {
...
-        if (requester.requestPostAnimationFrame) {
-            return requester.requestPostAnimationFrame(func);
-        } else if (requester.requestAnimationFrame) {
-            return requester.requestAnimationFrame(func);
-        } else {
-            return window.setTimeout(func, 16);
-        }
+        // fallback to the global `setTimeout`.
+        // In most cases (aka in the browser), `window` is the global object, so instead of calling `window.setTimeout` we could call the global `setTimeout`.
+        return setTimeout(func, 16) as unknown as number;
```
Comment (barroij) [inline packages/dev/core/src/Engines/thinEngine.ts:5882]: Here are the main changes of this PR.

We now test that the properties are of type `function` before calling them, which is safer but might be overkill.
And we do not call `window.setTimeout` anymore, we directly fallback to the global `setTimeout`.

If that's not ok, before the fallback I can add something like
```
if (typeof window.setTimeout === 'function') {
    return window.setTimeout(func);
}
```

File: packages/dev/inspector/src/components/actionTabs/tabs/propertyGrids/materials/textures/textureEditorComponent.tsx (hunk 2 above)
Comment (RaananW) [reply packages/dev/inspector/src/components/actionTabs/tabs/propertyGrids/materials/textures/textureEditorComponent.tsx:203]: why not use the global setTimeout and clearTimeout?

**Review by RaananW — APPROVED** [review]
LGTM! I am not sure why you opted in to use `window.clearTimeout`  instead of removing all `window` references in one place, but otherwise - this is a great change.

**Comment by RaananW** [comment]
And another possible change would be to use `ReturnType<setTimeout>` instead of casting to unknown and then to number, but that might collide with our internal type definitions in the engine. If that the case - leave it this way

File: packages/dev/inspector/src/components/actionTabs/tabs/propertyGrids/materials/textures/textureEditorComponent.tsx (hunk 2 above)
Comment (barroij) [reply packages/dev/inspector/src/components/actionTabs/tabs/propertyGrids/materials/textures/textureEditorComponent.tsx:203]: No good reason, I can change that

**Comment by barroij** [comment]
> And another possible change would be to use `ReturnType<setTimeout>` instead of casting to unknown and then to number, but that might collide with our internal type definitions in the engine. If that the case - leave it this way


Well the difficulty here is that actually `setTimeout` is used as a fallback in case `requestAnimationFrame`/`requestPostAnimationFrame` is not defined. So the return type of `QueueNewFrame` should the same as `requestAnimationFrame`, `requestPostAnimationFrame` and `setTimeout`...

**Comment by RaananW** [comment]
> 

let's leave it this way then.

File: packages/dev/inspector/src/components/actionTabs/tabs/propertyGrids/materials/textures/textureEditorComponent.tsx (hunk 2 above)
Comment (barroij) [reply packages/dev/inspector/src/components/actionTabs/tabs/propertyGrids/materials/textures/textureEditorComponent.tsx:203]: Actually in this case I can use window's function because we are in a React component.

In order to use global `setTimeout`/`clearTimeout `, I'd have to replace 
```
let throttleTimerId = -1
```
with 
```
let throttleTimerId : ReturnType<typeof setTimeout> | undefined = undefined;
```

and then I could write 
```
            clearTimeout(throttleTimerId);
            throttleTimerId = setTimeout(() => {
                this.raiseOnPropertyChanged(value, store);
            }, this.props.throttlePropertyChangedNotificationDelay ?? 200);
```

Comment (barroij) [reply packages/dev/inspector/src/components/actionTabs/tabs/propertyGrids/materials/textures/textureEditorComponent.tsx:203]: Shall I do it anyway ?

Comment (RaananW) [reply packages/dev/inspector/src/components/actionTabs/tabs/propertyGrids/materials/textures/textureEditorComponent.tsx:203]: nope, I totally missed the fact that this is a tsx and not in core. keep it this way :-)

**Comment by sebavan** [comment]
@bghgary could you check it would be ok with Native ???

**Comment by sebavan** [comment]
@barroij would be great if you can fix lint and format ?

**Comment by barroij** [comment]
@sebavan formatting done !

**Comment by RaananW** [comment]
I'll wait for @bghgary before merging just to be sure.
